### PR TITLE
277 chore periodic maintenance benchmarking and performance optimization review

### LINKED ISSUE #123: Implement maximum entropy deconvolution
### 🚶➡️🏃 Proposed Enhancements
In some spectroscopic fields like UV/Vis- or MIR-spectroscopy (of liquid systems), peaks can show very strong overlap. This limits the usefulness of many spectroscopic analysis techniques, e.g., Multivariate Fitting with reference spectra.
Having the peaks more resolved while keeping the noise suppressed would be a nice additional pre-processing step. Actually, using derivative spectroscopy is only a workaround to achieve just this.
This is easily depicted, e.g., looking at the derivaties of the MIR-spectra of some enzymes for protein analysis taken from 

[Baldassarre, et al., *Simultaneous Fitting of Absorption Spectra and Their Second Derivatives for an Improved Analysis of Protein Infrared Spectra*, Molecules 2015, 20(7), 12599-12622](https://doi.org/10.3390/molecules200712599)

image

The second and fourth order derivative reveal the overlapped peaks, but that's not easily achieved in practice where noise limits the usefulness of derivation.

However, in [Lórenz-Fonfriá & Padrós, *Maximum Entropy Deconvolution of Infrared Spectra: Use of a Novel Entropy Expression Without Sign Restriction*, Applied Spectroscopy, 2005, Volume 59, Number 4](https://www.researchgate.net/publication/7840935_Maximum_Entropy_Deconvolution_of_Infrared_Spectra_Use_of_a_Novel_Entropy_Expression_Without_Sign_Restriction), a quite powerful deconvolution technique based on Maximum Entropy Deconvolution was proposed. This can achieve such a peak resolution as well, but more resistant to noise. All in all, it circumvents the smoothing which would be mandatory as a pre-processing for taking derivatives.
From a resolution perspective, the results look promising (Figure A is sharpened to Figure C):

image

## 🧑‍💻 Implementation details
For this approach to work, weights have to provided (could be achieved with the functionality added for #44 and #120). 
The publication provides some implementation details on how to solve the underlying Nonlinear Optimization problem via a hand-crafted Conjugate Gradient method, but I think [`scipy.optimize.minimize`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html) offers more functionality to solve this in a graceful fashion:
- the problem incorporates a penalty weight $\lambda$ just like the Whittaker-Smoother. However, adjusting it to meet the reduced Chi-squared criterion proposed requires the evaluation of multiple $\lambda$-values which will be way more expensive than for the Whittaker-Smoother that has a straightforward linear solution. Reformulating the problem to *Maximize the entropy with the constraint that the Chi-squared criterion is roughly 1* should be more adequate and require only a single (but longer) optimization run.
- this would also allow to formulate the Jacobian and Hessian of the system in a more straigthforward way. Having this kind of gradient information is crucial for Nonlinear Optimization if we are aiming for speed. The Jacobian and the Hessian for the Chi-squared constraint can then be formulated as sparse matrices/linear operators that define the convolution operations in a very efficient way. On the other hand, the Jacobian and Hessian of the entropy terms can then be computed stand-alone rather than having the weighted Sum of Squared Residuals term to consider.

This is already a quite deep dive 🤿 into optimization theory, and I hope I can visualise 📊 it in a better way once the basic implementation is settled.

### PR DESCRIPTION
**Pull request by paucablop**
## Summary

- 

## Why is this change needed?

- 

## Closes

- 

## Related issues

- 

## Type of change

- [ ] Bug fix
- [ ] New feature
- [ ] Refactor
- [ ] Performance improvement
- [ ] Documentation update
- [ ] Tests only
- [ ] CI / DevOps / tooling
- [ ] Dependency update
- [ ] Breaking change

## What changed?

- 
- 
- 

## What did NOT change?

- 
- 

## API and compatibility impact

- [ ] No public API changes
- [ ] Public API added
- [ ] Public API changed
- [ ] Deprecation introduced
- [ ] Breaking change introduced

### Notes
- scikit-learn API compatibility:
- Affected modules/classes/functions:
- Backward compatibility considerations:

## Validation

- [ ] `task format-check`
- [ ] `task lint`
- [ ] `task spelling`
- [ ] `task type-check`
- [ ] `task test`
- [ ] `task coverage`
- [ ] `task test:matrix`
- [ ] `task docs:check`
- [ ] `task build`

### Validation notes

- 

## Tests

- [ ] Added new tests
- [ ] Updated existing tests
- [ ] No tests added because this PR only changes docs / CI / metadata

### Test coverage details
- Relevant test files:
- Edge cases covered:
- Numerical / estimator behavior checked:

## Documentation

- [ ] Docs updated
- [ ] Docstrings updated
- [ ] No docs update needed

### Documentation notes

- 

## Dependency / build / CI impact

- [ ] No dependency changes
- [ ] Runtime dependency change
- [ ] Dev dependency change
- [ ] GitHub Actions / CI changed
- [ ] Build/release process changed

### Notes
- 

## Reviewer focus

Please focus on:
- 
- 

## Screenshots / logs / benchmark output

<details>
<summary>Details</summary>
</details>

```text
Paste logs, benchmark numbers, or screenshots description here
```

## Checklist
- [ ] I linked the relevant issue(s) or explained why none exists
- [ ] I kept this PR scoped to a single purpose
- [ ] I added or updated tests where appropriate
- [ ] I updated documentation where appropriate
- [ ] I verified the change locally using the relevant tasks above
- [ ] I considered backward compatibility and public API impact
- [ ] I am ready for review

### CODE REVIEW

File: tests/smooth/test_whittaker_smooth.py [hunk 1]
```diff
@@ -60,3 +62,69 @@ def test_whittaker_smooth_banded_multi_row_matches_single_row():
...
+
+
+def test_whittaker_smooth_legacy_state_without_n_jobs():
+    # Arrange: simulate a pickle that pre-dates the n_jobs attribute
+    legacy = WhittakerSmooth()
+    legacy_state = {k: v for k, v in legacy.__dict__.items() if k != "n_jobs"}
+    restored = WhittakerSmooth()
+
+    # Act
+    restored.__setstate__(legacy_state)
```
[the remark this answers is not in the document]
Comment (paucablop) [reply tests/smooth/test_whittaker_smooth.py:128]: Fixed recreating a new object.

File: tests/baseline/test_as_ls.py [hunk 2]
```diff
@@ -40,3 +42,54 @@ def test_as_ls_sparse(spectrum, reference_asls):
...
+
+
+def test_as_ls_legacy_state_without_n_jobs():
+    # Arrange: simulate a pickle that pre-dates the n_jobs attribute
+    legacy = AsLs()
+    legacy_state = {k: v for k, v in legacy.__dict__.items() if k != "n_jobs"}
+    restored = AsLs()
+
+    # Act
+    restored.__setstate__(legacy_state)
```
[the remark this answers is not in the document]
Comment (paucablop) [reply tests/baseline/test_as_ls.py:93]: Fixed recreating a new object.

File: tests/baseline/test_air_pls.py [hunk 3]
```diff
@@ -40,3 +42,54 @@ def test_air_pls_sparse(spectrum, reference_airpls):
...
+
+
+def test_air_pls_legacy_state_without_n_jobs():
+    # Arrange: simulate a pickle that pre-dates the n_jobs attribute
+    legacy = AirPls()
+    legacy_state = {k: v for k, v in legacy.__dict__.items() if k != "n_jobs"}
+    restored = AirPls()
+
+    # Act
+    restored.__setstate__(legacy_state)
```
[the remark this answers is not in the document]
Comment (paucablop) [reply tests/baseline/test_air_pls.py:93]: Fixed recreating a new object.

File: tests/adaptation/test_x_axis_interpolator.py [hunk 4]
```diff
@@ -145,6 +146,47 @@ def test_parallel_matches_serial_for_per_row_x_axis(self, method):
...
+    def test_legacy_state_without_n_jobs(self):
+        """Ensures __setstate__ injects n_jobs=1 for pickles that pre-date the
+        attribute."""
+        # Arrange: simulate a pickle that pre-dates the n_jobs attribute
+        legacy = XAxisInterpolator(common_x_axis=np.linspace(0, 1, 5))
+        legacy_state = {k: v for k, v in legacy.__dict__.items() if k != "n_jobs"}
+        restored = XAxisInterpolator(common_x_axis=np.linspace(0, 1, 5))
+
+        # Act
+        restored.__setstate__(legacy_state)
```
[the remark this answers is not in the document]
Comment (paucablop) [reply tests/adaptation/test_x_axis_interpolator.py:184]: Fixed recreating a new object.

File: tests/baseline/test_rubberband_correction.py [hunk 5]
```diff
@@ -132,3 +134,54 @@ def test_rubberband_snapshot():
...
+
+
+def test_rubberband_legacy_state_without_n_jobs():
+    # Arrange: simulate a pickle that pre-dates the n_jobs attribute
+    legacy = RubberbandCorrection()
+    legacy_state = {k: v for k, v in legacy.__dict__.items() if k != "n_jobs"}
+    restored = RubberbandCorrection()
+
+    # Act
+    restored.__setstate__(legacy_state)
```
[the remark this answers is not in the document]
Comment (paucablop) [reply tests/baseline/test_rubberband_correction.py:185]: Fixed recreating a new object.

File: tests/baseline/test_ar_pls.py [hunk 6]
```diff
@@ -42,3 +44,54 @@ def test_ar_pls_sparse(spectrum_arpls, reference_arpls):
...
+
+
+def test_ar_pls_legacy_state_without_n_jobs():
+    # Arrange: simulate a pickle that pre-dates the n_jobs attribute
+    legacy = ArPls()
+    legacy_state = {k: v for k, v in legacy.__dict__.items() if k != "n_jobs"}
+    restored = ArPls()
+
+    # Act
+    restored.__setstate__(legacy_state)
```
[the remark this answers is not in the document]
Comment (paucablop) [reply tests/baseline/test_ar_pls.py:95]: Fixed recreating a new object.

File: tests/scatter/test_robust_normal_variate.py [hunk 7]
```diff
@@ -82,3 +83,54 @@ def test_rnv_percentile_above_100_rejected():
...
+
+
+def test_rnv_legacy_state_without_n_jobs():
+    # Arrange: simulate a pickle that pre-dates the n_jobs attribute
+    legacy = RobustNormalVariate()
+    legacy_state = {k: v for k, v in legacy.__dict__.items() if k != "n_jobs"}
+    restored = RobustNormalVariate()
+
+    # Act
+    restored.__setstate__(legacy_state)
```
[the remark this answers is not in the document]
Comment (paucablop) [reply tests/scatter/test_robust_normal_variate.py:134]: Fixed recreating a new object.
